packages/x-data-grid/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx

```diff
@@ -168,7 +168,7 @@ export const useGridColumnHeaders = (props: UseGridColumnHeadersProps) => {
       maxLastColumn = visibleColumns.length,
     } = params || {};
 
-    const firstColumnToRender = !hasVirtualization ? 0 : currentContext.firstColumnIndex;
```

Comment:
There was a bug here. `hasVirtualization` is using `gridVirtualizationColumnEnabledSelector` to get the column virtualization state. This was previously always `true`, even when the `disableVirtualization` prop was passed to DataGrid. The `0` here meant that pinned columns headers were duplicated were not being removed from the middle section of columns.